doc/developer/design/20231117_copy_to_s3.md

```diff
@@ -0,0 +1,279 @@
+```rust
+enum Uri {
+  // S3 path
+  S3(String),
+  ... // If we support more object stores or even local file, they can be added here
+}
+// and implement required traits so that it can be used as a valid value type
+impl TryFromValue<Value> for Uri {
+...
+}
+```
+To start with we can accept the S3 path as a string and in follow up PRs add the ability
+to eventually accept scalar expressions.
+TODO: Check if there's already some queries where we explicitly parse scalar expressions in the parser.
+
+#### `MAX FILE SIZE` Option
+Initially supporting `MAX FILE SIZE` as integer for the bytes will be pretty simple. Later we
+can introduce a `Memory` type which can take either an integer or a parseable string like `'2 GB'`,
+similar to the implementation for `Duration`.
+
+### Which cluster to run things on?
+Currently existing supported `COPY ... TO STDOUT` command fetches the result to `environmentd`.
+This will not be advisable for `COPY ... TO S3` as the data size could be large. Therefore,
+we should trigger a dataflow in a `clusterd` which should eventually write to S3 using
+user's aws connection details. The next question being, should it be a Storage cluser or a
+Compute cluster (at least till the time the cluster unification is done).
+
+We should use the Compute clsuter. Compute already has a concept of
+["Compute Sinks"](https://github.com/MaterializeInc/materialize/blob/v0.77.1/src/compute-types/src/sinks.rs#L100-L103),
+which include Subscribe and Persist (for materialized views). We can create another
+Compute S3 sink, which will actually read from persist and write to S3. Keeping this in a
+compute cluster also means,
+- We eventually will support `SELECT` queries in a `COPY ... TO S3` command and that can
+be a natural extension if things are already in compute.
+- Via the Persist Compute Sink, there's already precedence to talking to S3, we should
+be able to leverage some existing code.
+- This also will be consistent with the behaviour that a user can select the compute cluster
+where they want their query to run like they do for other queries.
+
+#### Other Alternative: Storage cluster
+The other alternative would have been to actually make use of a Storage cluster. If we think
+of `COPY ... TO S3` as a one time sink, then it might also make sense to have it in the Storage
+cluster. Especially if the initial scope does not include `SELECT` queries which
+would have required spinning up compute dataflows.
+Reasons to not do so,
+- As mentioned above, we will eventually support `SELECT` queries in `COPY ... TO S3`. The
+only reason it would have made sense to do this in storage would be to do as a quick stop gap
+if the effort is drastically less. It does not appear to be so.
+- Unlike Compute, Storage does not have a concept of running temporary/one-off dataflows.
+Everything which runs, like sources or sinks, do so continuously once scheduled till they are
+dropped.
+
+### Output file format
```

Comment:
Maybe for pretty simple datasets it should be fine. In general I don't think CSV can be lossless or at least it will need a lot of tweaks. It will also depend entirely upon what the options are when exporting and while importing. For example if somebody wants NULLs as empty string, it's changing the data type. Do we map strings to varchar or text?

Also, I need to check what happens with nested objects. I am assuming it will get serialized to some string, not sure if it will have enough information for it to be deserialized correctly.

Reply:
Btw, I just tried reimporting an exported nested type in postgres without overriding any option, it failed with invalid syntax because of nested double quotes. Will try later if it's possible to make it work with tweaks.

Reply:
Updated that it can be lossless if we escape correctly